Typing Indicator Animation

Rohitth007 opened the pull request:
This commit animates the typing feedback using an iterable cycle
object. This is done using an asynchronous function
`_show_typing_animation` and a boolean `is_recipient_typing` to
start and end the animation.

The corner case of not recieving a `stop` signal is handled by
setting `TYPING_STARTED_EXPIRY_PERIOD=15` and checking this
condition using `datetime`. 

**This has to be handled differently from #915 as footer keep updating every half second 
so, footer text's `duration` variable can't be used.**

Tests added.

![animate_typing](https://user-images.githubusercontent.com/64144419/114208939-af89c180-997b-11eb-90c8-0316f6076594.gif)

Rohitth007 committed:
bugfix: model/core: End typing notification on narrow change.

The `narrow` state in active_conversation_info is used so as to end
the notification if the user switches narrow by comparing with
`model.narrow`.

Tests added.

preetmishra commented on zulipterminal/model.py:
```diff
@@ -957,12 +973,10 @@ def _handle_typing_event(self, event: Event) -> None:
-                        ' is typing...'
-                    ])
+                    self.is_typing = True
+                    self.animate_typing(user)
```
I am not sure about the name `user` here. This person is actually a recipient and the ZT user is the _user_.

Rohitth007 replied:
This was how it originally was, though I think `sender` maybe a better word here?

preetmishra replied:
Hmm, `sender` would be a bit ambiguous as in this case both, the user and the recipient, could be perceived as a sender. I believe `recipient` would be apt.

preetmishra reviewed (changes requested):
@Rohitth007 Thanks for putting this forward! While this is a small addition, this does seem to improve the experience.

I have tested this locally and it seems to work well. :+1: 

Other than the in-line comments, there are a few things that be improved:
- We follow a particular convention for the commit messages. You can read the commit guidelines that we have in the README and also `$ git log --oneline`. I would recommend to you install `gitlint`.
- We would appreciate some tests to go along with the feature. Tests give us the ability to ship new features confidently.
- We use `mypy` for static type checking. Please add type annotation wherever needed.

Feel free to drop a message in **#zulip-terminal** if you need any assistance with anything.

Rohitth007 commented on zulipterminal/model.py:
```diff
@@ -945,6 +945,22 @@ def get_stream_by_id(streams: List[StreamData], stream_id: int
+    def animate_typing(self, user):
+        def typing_gen():
+            while True:
+                yield from ['', '.', '..', '...', '....']
```
Yes, even I put 3 initially but I added another just to see how it looks. Maybe 3 is good enough.

Rohitth007 commented on zulipterminal/model.py:
```diff
@@ -945,6 +945,22 @@ def get_stream_by_id(streams: List[StreamData], stream_id: int
 
+    @asynch
+    def animate_typing(self, user):
+        def typing_gen():
```
I will look into this :+1:

Rohitth007 commented on zulipterminal/model.py:
```diff
@@ -957,12 +973,10 @@ def _handle_typing_event(self, event: Event) -> None:
+                    self.is_typing = True
+                    self.animate_typing(user)
                 elif event['op'] == 'stop':
+                    self.is_typing = False
```
How about `sender_typing_status`?

preetmishra replied:
Suffixes like `status` convey that there are more than _two_ states that a variable can take. For booleans, it is reasonable to have prefixes like `is`, `has`, `should`. Perhaps, `is_recipient_typing`?

Rohitth007 commented:
@preetmishra Yes, I am aware of the things specified that I have not added yet. This is just a functional prototype and since it's small enough to understand, I was planning to write a proper commit structure after making those changes. I am in between my exams right now hence the not so perfect PR. Hope you understand :)

preetmishra commented:
@Rohitth007 No worries. Best of luck with the exams!

Rohitth007 commented on zulipterminal/model.py:
```diff
@@ -957,12 +973,10 @@ def _handle_typing_event(self, event: Event) -> None:
-                        ' is typing...'
-                    ])
+                    self.is_typing = True
+                    self.animate_typing(user)
```
Maybe `show_typing_animation` because there should be some way to indicate that it not just a normal screen update?

Rohitth007 commented:
@preetmishra PR updated for review

preetmishra commented on zulipterminal/model.py:
```diff
@@ -945,24 +947,44 @@ def get_stream_by_id(streams: List[StreamData], stream_id: int
+        start_time = datetime.now()
+        # Restore the footer if there hasn't been a corresponding typing
+        # 'stop' event after a 'start' event for greater than 15 seconds.
+        max_duration = timedelta(seconds=TYPING_STARTED_EXPIRY_PERIOD)
```
Has this constant (15 seconds) been borrowed from somewhere?

Rohitth007 replied:
Yes this is the same constant used in @prah23's #915 where I have borrowed it from and also the same is recommended in the API docs for typing events given [here](https://zulip.com/api/set-typing-status).

preetmishra commented on zulipterminal/model.py:
```diff
@@ -945,24 +947,44 @@ def get_stream_by_id(streams: List[StreamData], stream_id: int
+        # 'stop' event after a 'start' event for greater than 15 seconds.
+        max_duration = timedelta(seconds=TYPING_STARTED_EXPIRY_PERIOD)
+
+        for dots in itertools.cycle(['', '.', '..', '...']):
```
Could we use the `next` function with a `lambda` to iterate over the dots? That'd be a bit cleaner.

Rohitth007 replied:
I don't quite get this, next can be used only when a generator object is created right? Is lambda used to create this object? That would leads to more lines of code for getting the same functionality?

preetmishra replied:
Agreed; it would require more lines. However, it would facilitate is the separation of concerns. @neiljp Thoughts on this?

neiljp replied:
@preetmishra I'd prefer to see the comparison before judging.

Rohitth007 replied:
@preetmishra I've re-implemented this part with minor changes. I've not used lambda as I was getting an error but I think this is pretty much what you had in mind. It looks much more better and readable now though its the same number of lines because of the if statement coming out of the loop.

preetmishra commented on zulipterminal/model.py:
```diff
@@ -945,24 +947,44 @@ def get_stream_by_id(streams: List[StreamData], stream_id: int
+        for dots in itertools.cycle(['', '.', '..', '...']):
+            if not self.is_recipient_typing:
+                break
+            elif datetime.now() - start_time > max_duration:
```
This means that even if the recipient is still typing post 15 seconds, the user won't get any feedback, right?

Rohitth007 replied:
This means if the recipient hasn't sent a `stop` (which the senders client regularly sends if the user is idle or escaped out of compose box) for some reason like client crash etc, ZT keeps showing `typing...` .
If the recipient still is typing its the clients responsibility to tell ZT `start` and it will. So, ZT re-displays `typing...`

Rohitth007 replied:
Though I see now that I haven't reset `is_recipient_typing` to `False`. I think this doesn't change the logic but should I reset anyway?

preetmishra replied:
My concern is that we do not check whether we have a new `op=start` and just break regardlessly post `TYPING_STARTED_EXPIRY_PERIOD`. What would happen if we have another `op=start` event post `TYPING_STARTED_WAIT_PERIOD`?

Rohitth007 replied:
oooh! I haven't considered the case where a `start` is received while the async function is running resulting in 2 of them running parallely. Maybe a boolean to fix this?
And when a new `start` is received after 15 seconds this process can restart. So I don't see an issue while breaking after 15 sec, did I miss something?

Rohitth007 replied:
I have used the boolean `is_typing_animation_in_progress` to handle the case where 2 of them run in parallel.
It is a bit long but I didn't want to put `animation_in_progress` or `is_animating` as it doesn't tell us about what is animating.
Even `typing_in_progress` can seem similar to `is_recipient_typing`?

Rohitth007 replied:
@preetmishra If you could look at this particular discussion again I would mark it as resolved.

preetmishra reviewed (commented):
@Rohitth007 Hey, thanks for your work! I have given quick feedback on the approach. Let's discuss it before moving ahead for another review iteration.

Rohitth007 commented:
@preetmishra I have made a few more changes to handle corner cases better. Even the tests work better now. (I have moved time forward instead of reducing it)
It all works but PEP8 tells me that there is a trailing whitespace error on a line I haven't even touched.
I even tried using VSCode's `Trim Trailing Whitespace` command.
`zulipterminal/model.py:960:68: W291 trailing whitespace`

Rohitth007 commented:
@zulipbot add "PR needs review"
@zulipbot remove "PR awaiting update"

zee-bit commented:
> It all works but PEP8 tells me that there is a trailing whitespace error on a line I haven't even touched.

You did change that line. You can take a look at the `Files changed` tab of this commit and notice that line `960` is changed. You may be in a subsequent commit while running the linter and thought that you didn't change that line. :sweat_smile:
(That's why it's important to run tests for each commit so that you find your mistakes in that commit itself.)

**PRO-TIP:** `git blame` can prove really helpful in finding out who changed a particular line in a repository.

Rohitth007 commented:
Oh wait I was checking in a different file :sweat_smile: , I mistook `model.py` for `test_model.py` cause that was the only file I was editing at that time. :facepalm:

Ezio-Sarthak reviewed (commented):
Thanks for continuing work on this! Leaving some minor code-wise comments :+1:

The commit title could be improved a bit, something more specific?

Rohitth007 commented on zulipterminal/model.py:
```diff
@@ -1235,7 +1265,7 @@ def _handle_update_message_flags_event(self, event: Event) -> None:
         self, timestamp: int, *, show_seconds: bool, show_year: bool = False
     ) -> str:
-        local_time = datetime.datetime.fromtimestamp(timestamp)
+        local_time = datetime.fromtimestamp(timestamp)
```
Done this :+1:, makes sense to have it as a prep commit.

Rohitth007 commented on tests/model/test_model.py:
```diff
@@ -1660,83 +1661,139 @@ def test_toggle_stream_pinned_status(self, mocker, model,
-    ], ids=['not_in_pm_narrow', 'not_in_pm_narrow_with_sender',
-            'start', 'stop'])
+    @pytest.mark.parametrize(
+        'narrow, event, animation_in_progress, expect_recipient_typing', [
```
It should be interpreted as so:
```python
If "animation is in progress" and we "expect that the recipient is typing" then assert that this happens.
```
We don't expect animation status, maybe `is_animation_in_progress` is better?

Re `expected_recipient_typing_status` can be used if `recipient_typing_status` is the variable output of the function. Also from above discussion with Preet, 'status' implied more than 2 values and not a boolean.

I wanted to use `is_recipient_typing` but its the same name as boolean. How about `is_recipent_actually_typing`?
It would then be interpreted as:
```python
If "animation is in progress" and "recipient is actually typing" then assert that animation happens, etc...
```

preetmishra commented on tests/model/test_model.py:
```diff
@@ -2033,10 +2034,10 @@ def test_toggle_stream_pinned_status(
 
     @pytest.mark.parametrize(
-        "narrow, event, called",
+        "narrow, event, is_animation_in_progress, is_recipient_actually_typing",
```
While you're at it, let's use cases (`from pytest import param as case`), so that it's evident that which test case pairs with which test id.

Rohitth007 replied:
:+1:

preetmishra commented on tests/model/test_model.py:
```diff
@@ -2079,20 +2098,73 @@ def test_toggle_stream_pinned_status(
+            "not_in_pm_narrow",
+            "not_in_pm_narrow_with_sender",
+            "start",
+            "start when animating",
```
Let's use underscores as delimiters.

Rohitth007 replied:
:+1: Might have missed it

preetmishra commented on tests/model/test_model.py:
```diff
@@ -2079,20 +2098,73 @@ def test_toggle_stream_pinned_status(
+        narrow,
+        event,
+        is_animation_in_progress,
+        is_recipient_actually_typing,
```
Let's use an `expected_` as prefix for params that are expected from the test result, so `is_recipient_actually_typing` -> `expected_is_recipient_typing`.

Rohitth007 replied:
:+1:

preetmishra commented on tests/model/test_model.py:
```diff
@@ -2079,20 +2098,73 @@ def test_toggle_stream_pinned_status(
-        assert model.controller.view.set_footer_text.called == called
+        assert model.is_recipient_typing == is_recipient_actually_typing
+        assert model._show_typing_animation.called == (
+            is_recipient_actually_typing and not is_animation_in_progress
```
It'd be cleaner to parametrize over `_show_typing_animation_is_called` than derive it using these two variables?

Rohitth007 replied:
I think it would become too many variables to read and too many True's and False's. I've tried extracting the inside expression as another boolean variable(`expect_indicator`) inside the function and it looks better.

preetmishra replied:
The former shouldn't be an issue. I would avoid writing expressions/evaluations in the test code and focus on just testing the method with a bunch of inputs.

preetmishra commented on tests/model/test_model.py:
```diff
@@ -2079,20 +2098,73 @@ def test_toggle_stream_pinned_status(
+        mocker.patch("zulipterminal.ui.View.set_footer_text")
+        recipient = {"full_name": "hamlet"}
+        model.is_recipient_typing = event["op"] == "start"
+        model.last_update_time = datetime.now() - timedelta(seconds=14.5)
```
Could we have a test for `last_updated_time` > 15?

Rohitth007 replied:
I think this is already tested as the async function breaks when last_update_time crosses 15

Rohitth007 replied:
And also when `is_recipient_typing` is False it never enters, so breaking the loop because of `is_recipient_typing` is also tested.

preetmishra replied:
Right, I think this is fine for now.

preetmishra reviewed (commented):
@Rohitth007 Thanks for the improvements! This seems to work well locally. :+1:

The implementation looks good! I have left a few in-line comments. We could have another test for `last_updated_time` case.

Rohitth007 commented on zulipterminal/model.py:
```diff
@@ -1,9 +1,9 @@
 from collections import OrderedDict, defaultdict
 from concurrent.futures import Future, ThreadPoolExecutor, wait
 from copy import deepcopy
+from datetime import datetime
```
:+1: I made some changes here but is the part where I mention "this is a prep commit" fine?

Rohitth007 replied:
The problem with refactoring `test_ui_tools` is that `pytest` and `mypy` would get confused between the mocked `date` and the actual `date`. So its better to keep it as `datetime.date`

preetmishra replied:
I don't think it should be not an issue if we get the naming correct.

preetmishra replied:
It's just that we should be consistent with the imports or other conventions throughout the source. :)

Rohitth007 replied:
@preetmishra I was getting errors as it was using the patched `date` as the imported `date`.
I think you meant "don't think it should be an issue" or otherwise?

preetmishra replied:
You could change the variable name to differentiate between the two.

Yeah, that's what I meant. :) (I have amended my comment.)

Rohitth007 commented on tests/model/test_model.py:
```diff
@@ -1,6 +1,7 @@
 import json
 from collections import OrderedDict
 from copy import deepcopy
+from datetime import datetime, timedelta
```
I think the title is fine right? `ui` is the `area` and `model` and `test_model` are files changed.

Rohitth007 commented on zulipterminal/model.py:
```diff
@@ -993,11 +997,39 @@ def get_stream_by_id(streams: List[StreamData], stream_id: int) -> StreamData:
+        self.is_typing_animation_in_progress = True
+        dots = itertools.cycle(["   ", ".  ", ".. ", "..."])
+
+        # Untill `stop` signal is sent or
```
:sweat_smile: I'll fix it

Rohitth007 commented on zulipterminal/model.py:
```diff
@@ -1007,12 +1039,13 @@ def _handle_typing_event(self, event: Event) -> None:
-                    )
+                    recipient = self.user_dict[event["sender"]["email"]]
+                    self.is_recipient_typing = True
+                    self.last_update_time: datetime = datetime.now()
```
Is a name change to `typing_event_start_time` more clear instead of commenting?

preetmishra replied:
Yes! We should always improve the name instead of leaving comments in any case.

Rohitth007 commented:
@preetmishra This is ready for another review.

preetmishra commented on tests/model/test_model.py:
```diff
@@ -2079,20 +2095,66 @@ def test_toggle_stream_pinned_status(
         model.user_dict = {"hamlet@zulip.com": {"full_name": "hamlet"}}
+        model.is_recipient_typing = False
+        model.is_typing_animation_in_progress = is_animation_in_progress
+        expect_indicator = expected_is_recipient_typing and not is_animation_in_progress
```
It would be more explicit and readable if we put it as a parameter in the parametrization. Also, let's include `_show_typing_animation` in the name as with this we're essentially testing whether `_show_typing_animation` was called.

preetmishra commented on tests/model/test_model.py:
```diff
@@ -2079,20 +2095,66 @@ def test_toggle_stream_pinned_status(
+        assert model._show_typing_animation.called == expect_indicator
+
+    @pytest.mark.parametrize(
+        "event, expected_indicator",
```
`indicator` doesn't provide any context. Let's refer to it with the method name (see my other comment).

preetmishra commented on tests/model/test_model.py:
```diff
@@ -2079,20 +2095,66 @@ def test_toggle_stream_pinned_status(
+    def test__show_typing_animation(self, mocker, model, event, expected_indicator):
+        mocker.patch("zulipterminal.ui.View.set_footer_text")
+        recipient = {"full_name": "hamlet"}
+        model.is_recipient_typing = event["op"] == "start"
```
This is another candidate for a `param` in the parametrization block.

preetmishra reviewed (commented):
@Rohitth007 Thanks for persisting. This continues to work well and the PR looks good overall! :+1: I have left a few minor in-line comments.

You seem to have updated `last_update_time` in the code but it is still there in the second commit's description.

Rohitth007 commented:
@preetmishra This is ready for another review. I have made all the suggested changes except the datetime refactor for
test_ui_tools. Hopefully the rest of it is looking good?

preetmishra commented on tests/model/test_model.py:
```diff
@@ -2033,12 +2034,13 @@ def test_toggle_stream_pinned_status(
 
     @pytest.mark.parametrize(
-        "narrow, event, called",
+        (
```
Let's update `last_update_time` to `typing_event_start_time` in the commit description as I mentioned in the last review. :)

preetmishra commented on tests/model/test_model.py:
```diff
@@ -2079,20 +2101,60 @@ def test_toggle_stream_pinned_status(
+        assert model._show_typing_animation.called == expected__show_typing_animation
+
+    @pytest.mark.parametrize(
+        "event, expect_set_footer_animation",
```
minor: expected*

Rohitth007 replied:
I think expect is a better word here. Expected would we apt when we compare a variable with its expected value, but here every possibility makes a call to set_footer_text so I've used the above which would translate to - `if you expect footer animation to be set then do so and so tests else do something else`

Rohitth007 replied:
Maybe `expect_footer_animation` would be better?

preetmishra replied:
Alternatively, you could call it `is_set_footer_animation_called`?

preetmishra reviewed (commented):
@Rohitth007 Thanks for the update! It continues to work well. :+1:

@neiljp Other than the minor changes that I have left in-line, this looks good from my end.

Rohitth007 commented:
@preetmishra I have updated the PR with all the suggestions. Hopefully this is now complete.

neiljp commented on zulipterminal/model.py:
```diff
@@ -996,11 +1000,39 @@ def get_stream_by_id(streams: List[StreamData], stream_id: int) -> StreamData:
                             subscribers.remove(user_id)
 
+    @asynch
+    def _show_typing_animation(self, recipient: Dict[str, Any]) -> None:
```
This isn't just showing a typing animation, it's now the entire typing notification function.

Rohitth007 replied:
Yes one could say that, but it is mainly updating the animation "frame" by "frame" and deciding when to end it.
If future animations are done then  `create_animation` or similar function could be extracted out of this. For now I think calling it show_typing_notification and moving it to Controller would be better.

neiljp replied:
Your suggestion was along the lines of what I was meaning :+1:

neiljp reviewed (commented):
@Rohitth007 This seems to work well :+1:

Some thoughts:
* Does this incidentally resolve one issue which @prah23 was trying to fix in #915? (maybe we can all discuss) 
* Having the typing notification behavior extracted into a function is great - and necessary for the asynch stateful behavior of the animation :+1: (though as noted, this is not just animation handled there, but all the notification(s)!)
* While we have other code calling the controller/view from the model (including the current notification code), I'd prefer to move to a situation where we have that behavior wrapped in eg. the controller (maybe including the view?) - consider eg. the report_* functions.
* Related to that, the extra state you're storing in the model is really UI state - it feels strange having storage of typing animation state being in the model?
* Do we need to provide the full recipient data each time, or just the name?

Other than helping us move to a cleaner structure as we add new features, I'm hoping doing that with this PR will be helpful for handling multi-user typing and stream-typing events. Let me know if you want more specific pointers - we can certainly discuss :+1:

Rohitth007 commented:
@neiljp Thanks for the review. To address some of your points:
* Yes this also happens to resolve the same edge case as @prah23's #915 , where the basic idea was borrowed from, as it had to be handled differently as this is asynchronous.
* Moving this whole function into controller makes sense and would simplify work on typing notifications features later.
* The full recipient was not needed, it was used it the same way before hence I kept it. Maybe just sending the name is better.

prah23 commented on tests/model/test_model.py:
```diff
@@ -2079,20 +2101,60 @@ def test_toggle_stream_pinned_status(
         event["type"] = "typing"
 
-        mocker.patch("zulipterminal.ui.View.set_footer_text")
+        mocker.patch("zulipterminal.model.Model._show_typing_animation")
```
You can just use `MODEL + "._show_typing_animation"` here. The `MODEL` constant is defined at the top of the file as `zulipterminal.model.Model`. This will keep your patch consistent with the other tests in the file.

prah23 commented on tests/model/test_model.py:
```diff
@@ -2079,20 +2101,60 @@ def test_toggle_stream_pinned_status(
+    def test__show_typing_animation(
+        self, mocker, model, event, is_set_footer_animation_called
+    ):
+        mocker.patch("zulipterminal.ui.View.set_footer_text")
```
Echoing from the previous constant point, you can add a constant for `zulipterminal.ui.View` at the top of the file and use that here.

prah23 commented on tests/model/test_model.py:
```diff
@@ -2079,20 +2101,60 @@ def test_toggle_stream_pinned_status(
+        model._show_typing_animation(recipient)
+
+        if is_set_footer_animation_called:
+            assert model.controller.view.set_footer_text.called is True
```
I understand the general benefit of erring towards being verbose, but I think this statement will maintain readability even without the `is True` part of it. We can just
```python
assert that something is called
```
instead of 
```python
assert that something is called is True
```
The other benefit is consistency with the `assert`s in the other tests.

prah23 commented on tests/model/test_model.py:
```diff
@@ -2079,20 +2101,60 @@ def test_toggle_stream_pinned_status(
+            assert model.controller.view.set_footer_text.called is True
+            assert model.controller.view.set_footer_text.called_with()
+        else:
+            assert model.controller.view.set_footer_text.called_once_with()
```
This can instead use `assert_called_once_with()`? The assert a couple of lines above can also use `assert_called_with()`. The benefit, again, lies in consistency with the other tests.

prah23 commented on tests/model/test_model.py:
```diff
@@ -2079,20 +2101,60 @@ def test_toggle_stream_pinned_status(
+
+        if is_set_footer_animation_called:
+            assert model.controller.view.set_footer_text.called is True
+            assert model.controller.view.set_footer_text.called_with()
```
Can you also assert on what text the function is being called with? I understand that the dots might be tricky to assert on, but you could make use of `assert_has_calls`. If not with the dots, an assert with the base text?

Rohitth007 replied:
Thanks for this, it worked wonderfully

prah23 reviewed (changes requested):
Thanks for all the work @Rohitth007!
This PR looks good to me, and the expiry period condition seems to have been met. I have a few suggestions in terms of the tests that I'm listing in-line.

Rohitth007 commented:
I have updated this PR as per suggestions and also have improved the tests.

Rohitth007 commented:
Updated the PR as per discussions in CZO

Rohitth007 committed:
bugfix: model/core: End typing notification on unresponsive event.

The corner case of not receiving a stop signal is handled by
setting TYPING_STARTED_EXPIRY_PERIOD=15 and checking this
condition using datetime and comparing it with
`typing_start_time` state of `active_conversation_info`.

`typing_start_time` is used so that we don't miss a start
event (which is received every 10 seconds) and misinterpret it as
stop not received.

`active_conversation_info` is stored in a variable so that it
doesn't change dynamically while inside the loop.

Tests added.

neiljp reviewed (commented):
@Rohitth007 Thanks for the split - this becomes clearer now :+1:

Some of my inline feedback reflects some points I made previously, including regarding storing state in an appropriate namespace.

Rohitth007 commented on zulipterminal/model.py:
```diff
@@ -1130,12 +1130,14 @@ def _handle_typing_event(self, event: Event) -> None:
-                )
+                self.sender_name = self.user_dict[sender_email]["full_name"]
+                self.is_sender_typing = True
+                if not controller.is_typing_animation_in_progress:
```
Yes, makes sense I'll change it.

Rohitth007 commented on zulipterminal/core.py:
```diff
@@ -384,6 +387,25 @@ def open_in_browser(self, url: str) -> None:
+        while self.model.is_sender_typing:
+            self.view.set_footer_text(
+                [
+                    ("footer_contrast", " " + self.model.sender_name + " "),
+                    " is typing" + next(dots),
```
Making this as a refactor then animation commit makes sense but the problem is I cannot test only animation without a stop mode if the threads execute linearly

Rohitth007 commented on tests/model/test_model.py:
```diff
@@ -2304,21 +2341,44 @@ def test_toggle_stream_visual_notifications(
         model.narrow = narrow
         model.user_dict = {"hamlet@zulip.com": {"full_name": "hamlet"}}
         model.user_id = 5  # Iago's user_id
+        model.is_sender_typing = False
```
It does because it is outside the if conditional.

Rohitth007 commented on zulipterminal/model.py:
```diff
@@ -1130,12 +1130,14 @@ def _handle_typing_event(self, event: Event) -> None:
+
             elif event["op"] == "stop":
-                controller.view.set_footer_text()
+                self.is_sender_typing = False
```
I've used `active_conversation_info` dict for this which also makes it easier to implement group typing notification

neiljp commented:
@Rohitth007 Thanks for working on this different architecture - the first two commits of this are now merged :tada: 
(since the tests seem OK that far through)

I split the first commit to make it smaller and easier to read.
